File: sum-root-to-leaf-numbers.js
// Time Complexity : O(N)
// Space Complexity : O(H) where H is the height of the tree
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No, did this in class

/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */

// DFS
// var sumNumbers = function(root) {
//     return helper(root, 0);
// };

// const helper = (root, val1) => {
//     // Base Case
//     if (root == null) return 0;
//     if (root.left == null && root.right == null) return 10 * val1 + root.val;
    
//     // Logic
//     return helper(root.left, 10 * val1 + root.val) + helper(root.right, 10 * val1 + root.val)
// }

// BFS
var sumNumbers = function (root) {
    if (root == null) return 0;
    
    let sum = 0;
    let value = 0;
    let queue = [];
    queue.push([root, 0]);

    while (queue.length > 0) {
        let pair = queue.shift();
        let node = pair[0];
        let val1 = pair[1];
        
        if (node != null) {
            value = val1 * 10 + node.val;
            if (node.left == null && node.right == null) {
                sum += value;
            }
            queue.push([node.left, value]);
            queue.push([node.right, value]);
        }
    }
    return sum;
}

File: Sum _Root_to_LeafNumbers.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int result;
    public int sumNumbers(TreeNode root) {
        helper(root,0);
        return result;
    }
    private void helper(TreeNode root, int currNum){
        //base
        if(root==null) return;
        //logic
        helper(root.left,currNum*10 + root.val);
        if(root.left==null && root.right == null){
            result += currNum *10 + root.val;
        }
        helper(root.right,currNum*10 + root.val);
    }
}